smart_open/tests/test_smart_open.py

```diff
@@ -200,93 +202,158 @@ def test_http_pass(self):
+
+    @patch_string_io
+    def test_read_bytes(self):
+        """Can we read bytes from a byte stream?"""
+        buffer = six.BytesIO(self.text.encode())
+        with smart_open.smart_open(buffer, 'rb') as sf:
+            data = sf.read()
+        self.assertEqual(data, self.text.encode())
+
+    @patch_string_io
+    def test_write_bytes(self):
+        """Can we write bytes to a byte stream?"""
+        buffer = six.BytesIO()
+        with smart_open.smart_open(buffer, 'wb') as sf:
+            sf.write(self.text.encode())
+            self.assertEqual(buffer.getvalue(), self.text.encode())
+
+    @unittest.skipIf(PY2, "Python 2 does not make a difference between str and bytes")
+    @patch_string_io
+    def test_read_str(self):
+        """Can we read strings from a text stream?"""
+        buffer = six.StringIO(self.text)
+        with smart_open.smart_open(buffer, 'r') as sf:
+            self.assertRaises(TypeError, sf.read)  # we expect binary mode
+
+    @unittest.skipIf(PY2, "Python 2 does not make a difference between str and bytes")
+    @patch_string_io
+    def test_write_str(self):
+        """Can we write strings to a text stream?"""
+        buffer = six.StringIO()
+        with smart_open.smart_open(buffer, 'w') as sf:
+            self.assertRaises(TypeError, sf.write, self.text)  # we expect binary mode
+
+    def test_read_str_from_bytes(self):
+        """Can we read strings from a byte stream?"""
+        buffer = six.BytesIO(self.text.encode())
+        with smart_open.smart_open(buffer, 'r') as sf:
+            data = sf.read()
+        self.assertEqual(data, self.text)
+
+    def test_write_str_to_bytes(self):
+        """Can we write strings to a byte stream?"""
+        buffer = six.BytesIO()
+        with smart_open.smart_open(buffer, 'w') as sf:
+            sf.write(self.text)
+            self.assertEqual(buffer.getvalue(), self.text.encode())
+
+    def test_name_read(self):
+        """Can we use the "name" attribute to decompress on the fly?"""
+        data = b'Hello, world!' * 1000
+        buffer = six.BytesIO(bz2.compress(data))
+        buffer.name = 'data.bz2'
+        with smart_open.smart_open(buffer, 'rb') as sf:
+            data = sf.read()
+        self.assertEqual(data, data)
+
+    def test_name_write(self):
+        """Can we use the "name" attribute to compress on the fly?"""
+        data = b'Hello, world!' * 1000
+        buffer = six.BytesIO()
+        buffer.name = 'data.bz2'
+        with smart_open.smart_open(buffer, 'wb') as sf:
+            sf.write(data)
+        self.assertEqual(bz2.decompress(buffer.getvalue()), data)
+
+    def test_open_side_effect(self):
+        """Does our trick with "name" attribute work with open()?"""
+        data = b'Hello, world!' * 1000
+        with tempfile.NamedTemporaryFile(prefix='smart_open_tests_', delete=False) as tmpf:
+            tmpf.write(data)
+        try:
+            with open(tmpf.name, 'rb') as openf:
```

Comment:
What are you actually testing here?

1. Open a file given its path as a string
2. Read its contents
3. Compare the read contents against known values

How is this different to other tests that do the same thing? Why do you need to create and delete a temporary file as part of the test? It looks unnecessary. Please let me know if I'm missing something here.

Reply:
I am missing `.bz2` suffix here, right.

Reply:
So you're testing that you can do I/O with transparent decompression to file.bz2?

Let's make that obvious from the test name and docstring. test_side_effect tells nothing to the reader.

Reply:
We already test in-memory with `test_name_read`. This is the test to ensure that the regular `open()` works, and it requires a real file.

Adding more info to the docstring.